packages/yarnpkg-core/sources/scriptUtils.ts

```diff
@@ -158,6 +158,10 @@ export async function makeScriptEnv({project, locator, binFolder, lifecycleScrip
 
     scriptEnv.npm_package_name = structUtils.stringifyIdent(locator);
     scriptEnv.npm_package_version = version;
+
+    if (packageLocation) {
```

Comment:
You should get the `packageLocation` from the `project` and `locator` instead.

Reply:
It looks like `project` and `locator` do not contain a real path of a project. That's why it is calculated elsewhere (by a linker).

Just to be sure I pulled all strings that contained `yarn-a` (the name of a test project I use) from `project`.

<details>
<summary>That's what I found:</summary>

It is an array of pairs `[name, value]`, where:
* `name` is an array of names to reach `value`
  * Sets are converted to arrays.
  * Maps are converted to dictionaries.
* `value` is a string that satisfies a condition. In this case, it should contain a `yarn-a` substring.

Example from below:

```json
[
  ["workspaces", "0", "manifest", "raw", "dependencies", "yarn-a"],
  "../yarn-a"
]
```

it means: `project.workspaces[0].manifest.raw.dependencies["yarn-a"] === "../yarn-a"`

The whole data:

```json
[
  [
    [
      "workspaces",
      "0",
      "dependencies",
      "7d02798597d7b882fa8ad8ea49d94a6e1f374bfc2a316e242a070915229982007ddd9e467313acd106e95e016e962eec93a40505e1b2e88a49ff02fe49350d63",
      "name"
    ],
    "yarn-a"
  ],
  [
    [
      "workspaces",
      "0",
      "dependencies",
      "7d02798597d7b882fa8ad8ea49d94a6e1f374bfc2a316e242a070915229982007ddd9e467313acd106e95e016e962eec93a40505e1b2e88a49ff02fe49350d63",
      "range"
    ],
    "file:../yarn-a::locator=temp%40workspace%3A."
  ],
  [
    [
      "workspaces",
      "0",
      "manifest",
      "dependencies",
      "7d02798597d7b882fa8ad8ea49d94a6e1f374bfc2a316e242a070915229982007ddd9e467313acd106e95e016e962eec93a40505e1b2e88a49ff02fe49350d63",
      "name"
    ],
    "yarn-a"
  ],
  [
    [
      "workspaces",
      "0",
      "manifest",
      "dependencies",
      "7d02798597d7b882fa8ad8ea49d94a6e1f374bfc2a316e242a070915229982007ddd9e467313acd106e95e016e962eec93a40505e1b2e88a49ff02fe49350d63",
      "range"
    ],
    "../yarn-a"
  ],
  [
    ["workspaces", "0", "manifest", "raw", "dependencies", "yarn-a"],
    "../yarn-a"
  ],
  [
    [
      "storedPackages",
      "9a2edacdd76bdcc332243c868926236d82676b57fd00fda54f555f26817c13c868b68d3b31e37ba4659779a2c6c1f9121cdf5f2eaba391bcab4899e1163af28f",
      "name"
    ],
    "yarn-a"
  ],
  [
    [
      "storedPackages",
      "9a2edacdd76bdcc332243c868926236d82676b57fd00fda54f555f26817c13c868b68d3b31e37ba4659779a2c6c1f9121cdf5f2eaba391bcab4899e1163af28f",
      "reference"
    ],
    "file:../yarn-a#../yarn-a::hash=b0fe6c&locator=temp%40workspace%3A."
  ],
  [
    [
      "originalPackages",
      "9a2edacdd76bdcc332243c868926236d82676b57fd00fda54f555f26817c13c868b68d3b31e37ba4659779a2c6c1f9121cdf5f2eaba391bcab4899e1163af28f",
      "name"
    ],
    "yarn-a"
  ],
  [
    [
      "originalPackages",
      "9a2edacdd76bdcc332243c868926236d82676b57fd00fda54f555f26817c13c868b68d3b31e37ba4659779a2c6c1f9121cdf5f2eaba391bcab4899e1163af28f",
      "reference"
    ],
    "file:../yarn-a#../yarn-a::hash=b0fe6c&locator=temp%40workspace%3A."
  ]
]
```

</details>

The `yarn-a` project is actually in `.yarn/unplugged/yarn-a-file-9a2edacdd7`. As you can see the data above has no such string anywhere.

`packageLocation` is calculated in several places, e.g., here: https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-core/sources/scriptUtils.ts#L392

This is the relevant excerpt:

```ts
const configuration = project.configuration;

const linkers = project.configuration.getLinkers();
const linkerOptions = {project, report: new StreamReport({stdout: new PassThrough(), configuration})};

const linker = linkers.find(linker => linker.supportsPackage(pkg, linkerOptions));
if (!linker)
  throw new Error(`The package ${structUtils.prettyLocator(project.configuration, pkg)} isn't supported by any of the available linkers`);

const packageLocation = await linker.findPackageLocation(pkg, linkerOptions);
```

(No idea why there is an alias `const configuration = project.configuration;` yet in some places it still `project.configuration`).

Another such place is https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-core/sources/scriptUtils.ts#L494 It has the same algorithm and it calls `makeScriptEnv()`.

This algorithm involves a loop (`find()`) to find an appropriate linker, an asynchronous operation (`await linker.findPackageLocation()`), and `new StreamReport()`, which looks expensive, while I assume we want to keep `yarn` fast. That's why I decided to reuse that value by passing it to `makeScriptEnv()` rather than messing with linkers inside &mdash; I don't know what linkers do, I don't know what different implementations of linkers do, I don't know what future implementations of linkers would do.

So I need confirmation and sage advice:

* Do I need to get `packageLocation` separately rather than reusing a value by passing it in?
  * It is a judgment call based on performance considerations.
* Do we want to abstract away getting a package location? It looks like a function or a method would be appropriate in this case, which will encapsulate the snippet above.

Thank you in advance.

Reply:
@merceyz Obviously besides ^^^ (essentially, "or" choice) we can go with a hybrid solution: if `packageLocation` is not passed in as a parameter, we can calculate it using the possible function described above.

Reply:
See https://github.com/yarnpkg/berry/pull/4546#issuecomment-1181806355

> Do I need to get packageLocation separately rather than reusing a value by passing it in?

I don't want to risk users of the API not providing it and/or the value not matching up where the `Package` is located.

> Do we want to abstract away getting a package location?

Perhaps another time.